Shape-based initializers don't support fixed shapes

Not sure if this is supposed to work in the first place, but when trying to compile something like:
```cpp
auto a = xt::zeros<float>(xt::fixed_shape<64>{}); 
```

I get the following:

```cpp
Scanning dependencies of target xtensor_issue
[ 50%] Building CXX object CMakeFiles/xtensor_issue.dir/xtensor_issue.cpp.o
In file included from /Users/evgeny/dev/tmp/xtensor_issue.cpp:1:
In file included from /Users/evgeny/dev/xtensor/include/xtensor/xarray.hpp:13:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:1636:19: error: read-only variable is not assignable
        *__result = *__first;
        ~~~~~~~~~ ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:1661:19: note: in instantiation of function template specialization 'std::__1::__copy<const unsigned long *, const unsigned long *>' requested here
    return _VSTD::__copy(__unwrap_iter(__first), __unwrap_iter(__last), __unwrap_iter(__result));
                  ^
/Users/evgeny/dev/xtensor/include/xtensor/xbroadcast.hpp:286:14: note: in instantiation of function template specialization 'std::__1::copy<const unsigned long *, const unsigned long *>' requested here
        std::copy(s.begin(), s.end(), m_shape.begin());
             ^
/Users/evgeny/dev/xtensor/include/xtensor/xbroadcast.hpp:240:16: note: in instantiation of function template specialization 'xt::xbroadcast<xt::xscalar<float>, xt::fixed_shape<64> >::xbroadcast<float, xt::fixed_shape<64> >' requested here
        return broadcast_type(std::forward<E>(e), s);
               ^
In file included from /Users/evgeny/dev/tmp/xtensor_issue.cpp:1:
In file included from /Users/evgeny/dev/xtensor/include/xtensor/xarray.hpp:20:
In file included from /Users/evgeny/dev/xtensor/include/xtensor/xcontainer.hpp:21:
In file included from /Users/evgeny/dev/xtensor/include/xtensor/xaccessible.hpp:14:
/Users/evgeny/dev/xtensor/include/xtensor/xstrides.hpp:535:42: error: expression is not assignable
                output[output_index - 1] = static_cast<value_type>(input[input_index - 1]);
                ~~~~~~~~~~~~~~~~~~~~~~~~ ^
/Users/evgeny/dev/xtensor/include/xtensor/xbroadcast.hpp:287:13: note: in instantiation of function template specialization 'xt::broadcast_shape<std::__1::array<unsigned long, 0>, xt::fixed_shape<64> >' requested here
        xt::broadcast_shape(m_e.shape(), m_shape);
            ^
/Users/evgeny/dev/xtensor/include/xtensor/xbroadcast.hpp:240:16: note: in instantiation of function template specialization 'xt::xbroadcast<xt::xscalar<float>, xt::fixed_shape<64> >::xbroadcast<float, xt::fixed_shape<64> >' requested here
        return broadcast_type(std::forward<E>(e), s);
               ^
2 errors generated.
make[3]: *** [CMakeFiles/xtensor_issue.dir/xtensor_issue.cpp.o] Error 1
make[2]: *** [CMakeFiles/xtensor_issue.dir/all] Error 2
make[1]: *** [CMakeFiles/xtensor_issue.dir/rule] Error 2
make: *** [xtensor_issue] Error 2

```

--- a/include/xtensor/xbroadcast.hpp
+++ b/include/xtensor/xbroadcast.hpp
@@ -236,8 +236,9 @@ namespace xt
     template <class E, class S>
     inline auto broadcast(E&& e, const S& s)
     {
-        using broadcast_type = xbroadcast<const_xclosure_t<E>, S>;
-        return broadcast_type(std::forward<E>(e), s);
+        using shape_type = filter_fixed_shape_t<std::decay_t<S>>;
+        using broadcast_type = xbroadcast<const_xclosure_t<E>, shape_type>;
+        return broadcast_type(std::forward<E>(e), xtl::forward_sequence<shape_type, decltype(s)>(s));
     }
 
 #ifdef X_OLD_CLANG

--- a/include/xtensor/xshape.hpp
+++ b/include/xtensor/xshape.hpp
@@ -464,6 +464,33 @@ namespace xt
 
     template <class S>
     using index_from_shape_t = typename index_from_shape<S>::type;
+
+    /**********************
+     * filter_fixed_shape *
+     **********************/
+
+    namespace detail
+    {
+        template <class S>
+        struct filter_fixed_shape_impl
+        {
+            using type = S;
+        };
+
+        template <std::size_t... N>
+        struct filter_fixed_shape_impl<fixed_shape<N...>>
+        {
+            using type = std::array<std::size_t, sizeof...(N)>;
+        };
+    }
+
+    template <class S>
+    struct filter_fixed_shape : detail::filter_fixed_shape_impl<S>
+    {
+    };
+
+    template <class S>
+    using filter_fixed_shape_t = typename filter_fixed_shape<S>::type;
 }
 
 #endif

--- a/test/test_xbroadcast.cpp
+++ b/test/test_xbroadcast.cpp
@@ -38,6 +38,18 @@ namespace xt
         ASSERT_EQ(5.0, m1_assigned(0, 1, 1));
     }
 
+    TEST(xbroadcast, fixed_shape)
+    {
+        xarray<double> m = {{1, 2, 3}, {4, 5, 6}};
+        auto m_br = broadcast(m, xt::fixed_shape<1, 2, 3>());
+        ASSERT_EQ(1.0, m_br(0, 0, 0));
+        ASSERT_EQ(4.0, m_br(0, 1, 0));
+        ASSERT_EQ(5.0, m_br(0, 1, 1));
+        ASSERT_EQ(m_br.layout(), m.layout());
+        XT_EXPECT_ANY_THROW(m_br.at(0, 0, 0, 0));
+        XT_EXPECT_ANY_THROW(m_br.at(10, 10, 10));
+    }
+
     TEST(xbroadcast, element)
     {
         xarray<double> m1

